File: problems/28-fenwick-tree/003-2d-fenwick/java/Solution.java
public class Solution {
    private int rows, cols;
    private long[][] bit;

    public Solution(int rows, int cols) {
        this.rows = rows; this.cols = cols;
        bit = new long[rows + 1][cols + 1];
    }

    public void add(int row, int col, long delta) {
        for (int i = row + 1; i <= rows; i += i & -i)
            for (int j = col + 1; j <= cols; j += j & -j)
                bit[i][j] += delta;
    }

    public long prefix(int row, int col) {
        long sum = 0;
        for (int i = row + 1; i > 0; i -= i & -i)
            for (int j = col + 1; j > 0; j -= j & -j)
                sum += bit[i][j];
        return sum;
    }

    public long rangeSum(int r1, int c1, int r2, int c2) {
        return prefix(r2,c2)-prefix(r1-1,c2)-prefix(r2,c1-1)+prefix(r1-1,c1-1);
    }
}

File: problems/28-fenwick-tree/003-2d-fenwick/python/solution.py
class Fenwick2D:
    def __init__(self, rows, cols):
        self.rows, self.cols = rows, cols
        self.bit = [[0] * (cols + 1) for _ in range(rows + 1)]

    def add(self, r, c, delta):
        i = r + 1
        while i <= self.rows:
            j = c + 1
            while j <= self.cols:
                self.bit[i][j] += delta
                j += j & -j
            i += i & -i

    def prefix(self, r, c):
        total = 0
        i = r + 1
        while i > 0:
            j = c + 1
            while j > 0:
                total += self.bit[i][j]
                j -= j & -j
            i -= i & -i
        return total

    def range_sum(self, r1, c1, r2, c2):
        return (self.prefix(r2, c2) - self.prefix(r1 - 1, c2)
                - self.prefix(r2, c1 - 1) + self.prefix(r1 - 1, c1 - 1))
